module/owlib/src/c/ow_infernoembedded.c

```diff
@@ -0,0 +1,607 @@
+	if (device == NULL) {
+		return NULL;
+	}
+
+	memset(device->address, 0, sizeof(device->address));
+	device->device = UNKNOWN;
+	device->info = NULL;
+
+	return device;
+}
+
+/**
+ * A caller is done with a device (that has been malloced)
+ * If we have a hash caching the data, don't do anything, otherwise we leave dangling pointers in the hash
+ * If we don't have a hash, then free the device
+ */
+#if OW_UTHASH
+#define DEVICE_TERM(__dev)
+#else
+#define DEVICE_TERM(__dev) free_device(&__dev)
+#endif
+
+/**
+ * Free a device (and any device specific info)
+ * @param device a pointer to the device pointer, we will clear the pointer
+ */
+static void free_device(ie_device **device) {
+	owfree((*device)->info);
+	owfree(*device);
+	*device = NULL;
+}
+
+/**
+ * Invalidate a device (eg. if the device has been rebooted and will present as a different device)
+ * @param device a pointer to the device pointer
+ */
+static void invalidate_device(ie_device **device) {
+#if OW_UTHASH
+	HASH_DEL(device_cache, *device);
+#endif
+	free_device(device);
+}
+
+/**
+ * Convert a little endian byte array to a uint16_t
+ * @param buf the byte array to convert (must be at least 2 bytes long)
+ * @return the converted value
+ */
+static inline uint16_t read_uint16(BYTE *buf) {
+	uint16_t val = 0;
+
+	for (int i = 0; i < 2; i++) {
+		val >>= 8;
+		val |= buf[i] << 8;
+	}
+
+	return val;
+}
+
+/**
+ * Convert a little endian byte array to a uint32_t
+ * @param buf the byte array to convert (must be at least 4 bytes long)
+ * @return the converted value
+ */
+static inline uint32_t read_uint32(BYTE *buf) {
+	uint32_t val = 0;
+
+	for (int i = 0; i < 4; i++) {
+		val >>= 8;
+		val |= buf[i] << 24;
+	}
+
+	return val;
+}
+
+/**
+ * Convert a little endian byte array to a uint64_t
+ * @param buf the byte array to convert (must be at least 8 bytes long)
+ * @return the converted value
+ */
+static inline uint64_t read_uint64(BYTE *buf) {
+	uint64_t val = 0;
+
+	for (int i = 0; i < 8; i++) {
+		val >>= 8;
+		val |= ((uint64_t)buf[i]) << 56;
+	}
+
+	return val;
+}
+
+/**
+ * Convert a uint16_t to a little endian byte array
+ * @param val the value to convert
+ * @param buf the buffer to write to (must be at least 2 bytes long)
+ */
+static inline void write_uint16(uint16_t val, BYTE *data) {
+	data[0] = (val & 0x000000FF);
+	data[1] = (val & 0x0000FF00) >> 8;
+}
+
+/**
+ * Convert a uint32_t to a little endian byte array
+ * @param val the value to convert
+ * @param buf the buffer to write to (must be at least 4 bytes long)
+ */
+static inline void write_uint32(uint32_t val, BYTE *data) {
+	data[0] = (val & 0x000000FF);
+	data[1] = (val & 0x0000FF00) >> 8;
+	data[2] = (val & 0x00FF0000) >> 16;
+	data[3] = (val & 0xFF000000) >> 24;
+}
+
+/**
+ * Convert a uint64_t to a little endian byte array
+ * @param val the value to convert
+ * @param buf the buffer to write to (must be at least 8 bytes long)
+ */
+static inline void write_uint64(uint32_t val, BYTE *data) {
```

Comment:
why is `write_uint64` needed? It is never called

Reply:
Thanks, it's also buggy (the prototype takes in a uint32_t rather than uint64_t)